snuba/state/rate_limit.py

```diff
@@ -0,0 +1,164 @@
+from abc import ABC, abstractmethod
+from collections import namedtuple
+from contextlib import AbstractContextManager, ExitStack
+import logging
+import time
+import uuid
+
+from snuba import state, util
+
+logger = logging.getLogger('snuba.state.rate_limit')
+
+
+class RateLimit(AbstractContextManager, ABC):
+    """
+    A context manager for rate limiting that allows for limiting based on
+    on a rolling-window per-second rate as well as the number of requests
+    concurrently running.
+
+    Uses a single redis sorted set per rate-limiting bucket to track both the
+    concurrency and rate, the score is the query timestamp. Queries are thrown
+    ahead in time when they start so we can count them as concurrent, and
+    thrown back to their start time once they finish so we can count them
+    towards the historical rate.
+
+               time >>----->
+    +-----------------------------+--------------------------------+
+    | historical query window     | currently executing queries    |
+    +-----------------------------+--------------------------------+
+                                  ^
+                                 now
+    """
+
+    def __init__(self, request):
+        self._request = request
+        self.__did_run = False
```

Comment:
Not a big fan of this flag. 

Background: when we put queries into Redis, we put them in the future. So that when other queries come in, we can count them towards the current rate. Then, once the query has been run, we put it back to it's start time.

Sometimes, however, we don't put the query in Redis (for example, when we want to bypass the rate limit). In those cases, we do not need to put the query back to the start time. That's why this flag exists.

Reply:
Moving back to the function as context manager you would not need this flag.